### PR TITLE
Payments

### PR DESCRIPTION
**Pull request by kshitij-nawandar9**

### CODE REVIEW

File: internal/controllers/payments/controller.go [hunk 1]
```diff
@@ -5,19 +5,61 @@ import (
...
+	"strconv"
+	"time"
 )
 
-func AddPayment(c *gin.Context) {
+func AddPayments(c *gin.Context) {
 	var payment payments.Payments
 	err := c.BindJSON(&payment)
 	if err != nil {
+		fmt.Println(err.Error())
```
Comment (s3e3) [inline internal/controllers/payments/controller.go]: remove print

File: internal/controllers/payments/controller.go [hunk 2]
```diff
@@ -5,19 +5,61 @@ import (
...
 		return
```
Comment (s3e3) [inline internal/controllers/payments/controller.go:13]: return 400 error code

File: internal/controllers/payments/controller.go [hunk 3]
```diff
@@ -5,19 +5,61 @@ import (
...
 	}
+
 	err2 := payments.AddPayments(&payment)
 	if err2 != nil {
-		fmt.Println(err.Error())
-		c.AbortWithStatus(http.StatusNotFound)
+		c.String(http.StatusOK,"Payment failed")
```
Comment (s3e3) [inline internal/controllers/payments/controller.go]: check why payment failed and return proper error accordingly

File: internal/controllers/payments/controller.go [hunk 4]
```diff
@@ -5,19 +5,61 @@ import (
...
+		c.JSON(http.StatusOK, payment)
+	}
+}
+
+func GetPaymentByID(c *gin.Context) {
+	var payment payments.Payments
+	id := c.Params.ByName("payments_id")
+	err := payments.GetPaymentByID(&payment, id)
+	if err != nil {
+		c.String(http.StatusOK,"Record not found")
```
Comment (s3e3) [inline internal/controllers/payments/controller.go]: return http 404

File: internal/controllers/payments/controller.go [hunk 5]
```diff
@@ -5,19 +5,61 @@ import (
...
+	}
+}
+
+func GetPaymentsByTime(c *gin.Context) {
+	var payment []payments.Payments
+	from := c.Query("from")
+	to := c.Query("to")
+	var startTime,endTime int64
+	var err error
+	if from==""{
```
Comment (s3e3) [inline internal/controllers/payments/controller.go]: add proper spacing

File: internal/controllers/payments/controller.go [hunk 6]
```diff
@@ -5,19 +5,61 @@ import (
...
+		startTime=time.Now().Unix()
+	}else{
+		startTime,err=strconv.ParseInt(from,10,64)
+		if err!=nil{
+			return
+		}
+	}
+	if to==""{
```
Comment (s3e3) [inline internal/controllers/payments/controller.go]: add proper spacing

File: internal/controllers/payments/controller.go [hunk 7]
```diff
@@ -5,19 +5,61 @@ import (
...
+		endTime=time.Now().Unix()
+	}else{
+		endTime,err=strconv.ParseInt(to,10,64)
+	}
+	fmt.Println(startTime,endTime)
+	err2 := payments.GetPaymentsByTime(&payment, startTime, endTime)
+	if err2 != nil {
+		c.String(http.StatusOK,"Request failed.")
```
Comment (s3e3) [inline internal/controllers/payments/controller.go]: return 400/404 accordingly

File: internal/entities/payments/payments/core.go [hunk 8]
```diff
@@ -1,12 +1,92 @@
...
 
-func AddPayments(payments *Payments) (err error) {
-	if err = config.DB.Table("payments").Create(payments).Error; err != nil {
+var InputPaymentsChannel =make(chan *Payments,1000)
+var ResultsPaymentsChannel=make(chan *Payments,1000)
+
+func AddPayments(payment *Payments) (err error) {
+	payment.Type=GetPaymentType(payment)
+	payment.Status="processing"
+	payment.ID="paymt_"+utilities.RandomString(14)
```
Comment (s3e3) [inline internal/entities/payments/payments/core.go]: for every model there should be a small method like `generateId` that does this.
Also keep `paymt` in `payments/model.go` as const

File: internal/entities/payments/payments/core.go [hunk 9]
```diff
@@ -1,12 +1,92 @@
...
+	//err=ValidityCheck(payment)
+	//if err!=nil{
+	//	return err
+	//}
+	InputPaymentsChannel <-payment
```
Comment (s3e3) [inline internal/entities/payments/payments/core.go:23]: add validations:
- if user has enough balance to transfer money
- if bank accs. exist

File: internal/entities/payments/payments/core.go [hunk 10]
```diff
@@ -1,12 +1,92 @@
...
+	if err= config.DB.Table("payments").Create(payment).Error; err != nil {
+		return err
+	}
+	return nil
+}
+
+func GetPaymentByID(payment *Payments, id string) (err error) {
+	if err = config.DB.Table("payments").Where("id = ?", id).First(payment).Error; err != nil {
```
Comment (s3e3) [inline internal/entities/payments/payments/core.go]: create a `payments/repo.go` and add DB query there. core shouldn't directly talk to DB

File: internal/entities/payments/payments/core.go [hunk 11]
```diff
@@ -1,12 +1,92 @@
...
+		case payment:=<-ResultsPaymentsChannel:
+			err := UpdatePayment(payment)
+			if err != nil {
+				return
+			}
+		}
+	}
+}
+
+func GetUserID(searchID string) (err error,userID string) {
```
Comment (s3e3) [inline internal/entities/payments/payments/core.go:66]: why this method is in payments core?

File: internal/entities/payments/payments/core.go [hunk 12]
```diff
@@ -1,12 +1,92 @@
...
+	if err = config.DB.Table(table).Where("id = ?", searchID).First(userID).Error; err != nil {
+		return err,""
+	}
+	return nil,userID
+}
+
+func GetPaymentType(payment *Payments) string{
+	if strings.HasPrefix(payment.SourceId,"wallt"){
+		if strings.HasPrefix(payment.DestinationId,"wallt"){
+			return "wallet transfer"
```
Comment (s3e3) [inline internal/entities/payments/payments/core.go]: keep strings in constants

File: internal/entities/payments/payments/model.go [hunk 13]
```diff
@@ -1,19 +1,18 @@
 package payments
 
-import "gorm.io/gorm"
-
 type Payments struct {
-	gorm.Model
+	//gorm.Model
```
Comment (s3e3) [inline internal/entities/payments/payments/model.go:11]: why commented?

**Review by s3e3 — CHANGES_REQUESTED** [review]
some changes

File: internal/controllers/payments/controller.go (hunk 1 above)
Comment (kshitij-nawandar9) [reply internal/controllers/payments/controller.go]: resolved

File: internal/controllers/payments/controller.go (hunk 2 above)
Comment (kshitij-nawandar9) [reply internal/controllers/payments/controller.go:13]: resolved

File: internal/controllers/payments/controller.go (hunk 4 above)
Comment (kshitij-nawandar9) [reply internal/controllers/payments/controller.go]: resolved

File: internal/controllers/payments/controller.go (hunk 5 above)
Comment (kshitij-nawandar9) [reply internal/controllers/payments/controller.go]: resolved

File: internal/controllers/payments/controller.go (hunk 6 above)
Comment (kshitij-nawandar9) [reply internal/controllers/payments/controller.go]: resolved

File: internal/controllers/payments/controller.go (hunk 7 above)
Comment (kshitij-nawandar9) [reply internal/controllers/payments/controller.go]: resolved

File: internal/entities/payments/payments/core.go (hunk 8 above)
Comment (kshitij-nawandar9) [reply internal/entities/payments/payments/core.go]: resolved

File: internal/entities/payments/payments/core.go (hunk 12 above)
Comment (kshitij-nawandar9) [reply internal/entities/payments/payments/core.go]: resolved

**Pull request merged**